src/theory/bags/bag_reduction.h

```diff
@@ -0,0 +1,78 @@
+/******************************************************************************
+ * Top contributors (to current version):
+ *   Mudathir Mohamed
+ *
+ * This file is part of the cvc5 project.
+ *
+ * Copyright (c) 2009-2021 by the authors listed in the file AUTHORS
+ * in the top-level source directory and their institutional affiliations.
+ * All rights reserved.  See the file COPYING in the top-level source
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * bag reduction.
+ */
+
+#ifndef CVC5__BAG_REDUCTION_H
+#define CVC5__BAG_REDUCTION_H
+
+#include <vector>
+
+#include "cvc5_private.h"
+#include "smt/env.h"
+#include "theory/bags/inference_manager.h"
+
+namespace cvc5 {
+namespace theory {
+namespace bags {
+
+/**
+ * class for bag reductions
+ */
+class BagReduction
+{
+ public:
+  BagReduction(Env& env);
+  ~BagReduction();
+  /**
+   * @param node a term of the form (bag.fold f t A) where
+   *        f: (-> T1 T2 T2) is a binary operation
+   *        t: T2 is the initial value
+   *        A: (Bag T1) is a bag
+   * @param asserts a list of assertions generated by this reduction
+   * @return the reduction term (combine n) such that
+   * (and
+   *  (forall ((i Int))
+   *    (let ((iMinusOne (- i 1)))
+   *      (let ((uf_i (uf i)))
+   *        (=>
+   *          (and (>= i 1) (<= i n))
+   *          (and
+   *            (= (combine i) (f uf_i (combine iMinusOne)))
+   *            (=
+   *              (unionDisjoint i)
+   *              (bag.union_disjoint
+   *                (bag uf_i 1)
+   *                (unionDisjoint iMinusOne))))))))
+   *   (= (combine 0) t)
+   *   (= (unionDisjoint 0) (as bag.empty (Bag T1)))
+   *   (= A (unionDisjoint n))
+   *   (>= n 0))
+   *   where
+   *   n: Int is the cardinality of bag A
+   *   uf:Int -> T1 is an uninterpreted function that represents elements of A
+   *   combine: Int -> T2 is an uninterpreted function
+   *   unionDisjoint: Int -> (Bag T1) is an uninterpreted function
+   */
+  Node reduceFoldOperator(Node node, std::vector<Node>& asserts);
+
+ private:
+  /** The associated environment. */
+  Env& d_env;
```

Comment:
As a style point here, you should make this class inherit from EnvObj, and not have an explicit reference d_env.

Reply:
Done.